fuel-core/src/schema/contract.rs

```diff
@@ -49,3 +60,34 @@ impl ContractQuery {
         Ok(Some(contract))
     }
 }
+
+#[derive(Default)]
+pub struct ContractBalanceQuery;
+
+#[Object]
+impl ContractBalanceQuery {
+    async fn contract_balance(
+        &self,
+        ctx: &Context<'_>,
+        contract: ContractId,
+        asset: AssetId,
+    ) -> async_graphql::Result<U64> {
+        let contract_id: fuel_types::ContractId = contract.0;
+
+        let db = ctx.data_unchecked::<Database>().clone();
+
+        let asset_id: fuel_types::AssetId = asset.into();
+
+        let result = fuel_vm::storage::InterpreterStorage::merkle_contract_asset_id_balance(
+            &db,
+            &contract_id,
+            &asset_id,
+        );
```

Comment:
Nice. Even though I was very comfortable with GQL, I found it hard to find ways to reach the data I actually needed.

So this is how we get contract balances, huh? Is there a way to use this to build the plural API `Query.contractBalances`? Something to get all balances of a contract?

Reply:
I had was having trouble as well accessing the data until I found the [fuel-vm docs](https://docs.rs/fuel-core/0.6.3/fuel_core/database/struct.Database.html#method.merkle_contract_asset_id_balance) which are very helpful for fetching a data from the db.

I don't know of a specific way to fetch all the balances fro a contract, but my thinking was using just accepting a vec of assetId's for `Query.contractBalances` and looping through each of them when fetching balances.

Reply:
All tables are iterable, performing these kinds of queries depends on the key structure, as all k/v pairs are lexicographically sorted by their bytes. You can see that balance keys are inserted in [database/balances.rs](https://github.com/FuelLabs/fuel-core/blob/master/fuel-core/src/database/balances.rs#L24) as a (contract_id, asset_id) concatenated tuple. So for enumerating all the balances for a contract, you'd provide just the `contract_id` prefix to the query and the database could iterate over all the assets for that prefix. Operations such as sorting based on balance value would be more suited for fuel-indexer though.